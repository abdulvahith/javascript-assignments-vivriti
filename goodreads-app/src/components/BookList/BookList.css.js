export default {
  root: {
    flexGrow: 1,
    width: '100%'
  },
  container: {
    height: '90%',
    overFlow: "auto"
  }
}
