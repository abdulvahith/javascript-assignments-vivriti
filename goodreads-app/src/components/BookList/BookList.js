import React, { Component } from 'react';
import PropTypes from "prop-types";
import Paper from '@material-ui/core/Paper';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TablePagination from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';
import { fade, makeStyles } from '@material-ui/core/styles';
import Styles from "./BookList.css.js";
import Avatar from '@material-ui/core/Avatar';
import Link from '@material-ui/core/Link';

const columns = [
  { id: 'title', label: 'Book Name', minWidth: 170 },
  { id: 'small_image_url', label: 'Cover Image', minWidth: 100 , format: "img"},
  { id: 'authors', label: 'Authors', format: "join" },
  { id: 'description', label: 'Book Description', minWidth: 200, align: 'left', format: "html"},
  { id: 'link', label: 'Book Link', minWidth: 100, align: 'right', format: "link" },
];

const searchResultcolumns = [
  { id: 'title', label: 'Book Name', minWidth: 170 },
  { id: 'small_image_url', label: 'Cover Image', minWidth: 100 , format: "img"},
  { id: 'authors', label: 'Authors' },
];

export default class BookList extends Component {
  constructor(props) {
    super(props);
    this.state= {
      page: 0,
      rowsPerPage: 10
    }
    this.handleChangePage= this.handleChangePage.bind(this);
    this.handleChangeRowsPerPage= this.handleChangeRowsPerPage.bind(this);
    this.htmlDecode= this.htmlDecode.bind(this);
  }

  handleChangePage(event, newPage){
    this.setState({
      page: newPage
    })
  };

  handleChangeRowsPerPage(event){
    console.log("RowsPerPage", +event.target.value);
    this.setState({
      rowsPerPage: +event.target.value,
      page: 0
    })
  };

  htmlDecode(content) {
    let e = document.createElement('div');
    e.innerHTML = content;
    return e.childNodes.length === 0 ? "" : e.childNodes[0].nodeValue;
  }

  render() {
    let { page, rowsPerPage }= this.state;
    let { bookList, isSearch }=this.props;

    if (isSearch) {
      bookList= bookList && Array.isArray(bookList) && bookList.map((item)=>{
        return { title: item.best_book.title, small_image_url: item.best_book.small_image_url, authors: item.best_book.author.name }
      })
    }

    let headerColumn= isSearch ? searchResultcolumns : columns;

    if (!(bookList && Array.isArray(bookList))) {
      return <div> No results found  </div>
    }

    return (
      <Paper style={Styles.root}>
        <TableContainer style={Styles.container}>
          <Table stickyHeader aria-label="sticky table">
            <TableHead>
              <TableRow>
                {headerColumn.map(column => (
                  <TableCell
                    key={column.id}
                    align={column.align}
                    style={{ minWidth: column.minWidth }}
                  >
                    {column.label}
                  </TableCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {bookList.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage).map((row, index) => {
                return (
                  <TableRow hover role="checkbox" tabIndex={-1} key={row.code} key={index} >
                    {headerColumn.map(column => {
                      const value = row[column.id];
                      return (
                        <TableCell key={column.id} align={column.align}>
                          {
                            column.format==="img" ?
                            <Avatar alt="Remy Sharp" variant="square" src={value} sizes="50"/>:
                            (column.format=== "join"?
                              value.author.name :
                            (column.format==="html" ?
                            <div dangerouslySetInnerHTML={{ __html: value }} className="SearchResult-body"/>:
                              (column.format==="link" ?
                              <Link
                                component="button"
                                variant="body2"
                                onClick={() => {
                                  var win = window.open(value, '_blank');
                                  win.focus();
                                }}
                                >
                                Book Link
                              </Link>
                              :value)
                            )
                           )
                          }
                        </TableCell>
                      );
                    })}
                  </TableRow>
                );
              })}
            </TableBody>
          </Table>
        </TableContainer>
        <TablePagination
          rowsPerPageOptions={[10, 25, 100]}
          component="div"
          count={bookList.length}
          rowsPerPage={rowsPerPage}
          page={page}
          onChangePage={this.handleChangePage}
          onChangeRowsPerPage={this.handleChangeRowsPerPage}
        />
      </Paper>
    );
  }
}

BookList.propTypes = {
  bookList: PropTypes.array,
  isSearch: PropTypes.bool
}
