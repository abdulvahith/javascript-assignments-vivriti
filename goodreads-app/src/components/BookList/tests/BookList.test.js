import React from 'react';
import renderer from 'react-test-renderer';
import BookList from '../BookList';
import { mount, shallow } from 'enzyme';

describe('BookList component should render properly', () => {
  it('renders without crashing', () => {
    shallow(<BookList />);
  });

  it('renders correctly when issearch false snapshot test', () => {
     let bookList= [
       {
         title:"Tribe",
         small_image_url: "https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/books/1580496015l/50731895._SX50_.jpg",
         authors: {author:{name: "Timothy"} },
         description: "<div><b>We All Need Mentors. Here Are More than 100 of the Worldâ€™s Best.</b>",
         link: "https://www.goodreads.com/book/show/50731895-tribe-of-mentors"
       }
     ]
     const rendered = renderer.create(
       <BookList
         bookList= {bookList}
         isSearch= {false}
       />
     );
     expect(rendered.toJSON()).toMatchSnapshot();
   });

  it('renders correctly when issearch true snapshot test', () => {
    let bookList= [
      {
        best_book: {
          title:"Tribe",
          small_image_url: "https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/books/1580496015l/50731895._SX50_.jpg",
          author: {name: "Joelle Charbonneau"}
        }
      }
    ]
    const rendered = renderer.create(
      <BookList
        bookList= {bookList}
        isSearch= {true}
      />
    );
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('getBookList did mount get called', () => {
    const getBookList = jest.fn();

    const rendered = renderer.create(
      <BookList
        getBookList= {getBookList}
      />
    );

    setTimeout(() => {
      expect(getBookList).toHaveBeenCalled();
      }, 500);
  });
});
