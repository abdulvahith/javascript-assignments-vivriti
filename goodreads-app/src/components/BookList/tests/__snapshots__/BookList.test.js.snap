// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList component should render properly renders correctly when issearch false snapshot test 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "flexGrow": 1,
      "width": "100%",
    }
  }
>
  <div
    className="MuiTableContainer-root"
    style={
      Object {
        "height": "90%",
        "overFlow": "auto",
      }
    }
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader"
    >
      <thead
        className="MuiTableHead-root"
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 170,
              }
            }
          >
            Book Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            Cover Image
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": undefined,
              }
            }
          >
            Authors
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 200,
              }
            }
          >
            Book Description
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            Book Link
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
      >
        <tr
          className="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            Tribe
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            <div
              className="MuiAvatar-root MuiAvatar-square"
            >
              <img
                alt="Remy Sharp"
                className="MuiAvatar-img"
                sizes="50"
                src="https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/books/1580496015l/50731895._SX50_.jpg"
              />
            </div>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            Timothy
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
          >
            <div
              className="SearchResult-body"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<div><b>We All Need Mentors. Here Are More than 100 of the Worldâ€™s Best.</b>",
                }
              }
            />
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          >
            <button
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiLink-button MuiTypography-body2 MuiTypography-colorPrimary"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Book Link
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        className="MuiTablePagination-spacer"
      />
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby=" "
          className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          10
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value={10}
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        1-1 of 1
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Previous page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Next page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`BookList component should render properly renders correctly when issearch true snapshot test 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "flexGrow": 1,
      "width": "100%",
    }
  }
>
  <div
    className="MuiTableContainer-root"
    style={
      Object {
        "height": "90%",
        "overFlow": "auto",
      }
    }
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader"
    >
      <thead
        className="MuiTableHead-root"
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 170,
              }
            }
          >
            Book Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            Cover Image
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": undefined,
              }
            }
          >
            Authors
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
      >
        <tr
          className="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            Tribe
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            <div
              className="MuiAvatar-root MuiAvatar-square"
            >
              <img
                alt="Remy Sharp"
                className="MuiAvatar-img"
                sizes="50"
                src="https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/books/1580496015l/50731895._SX50_.jpg"
              />
            </div>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
          >
            Joelle Charbonneau
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        className="MuiTablePagination-spacer"
      />
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby=" "
          className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          10
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value={10}
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        1-1 of 1
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Previous page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Next page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
